File: iOS/Crisis Connect/Security/RoleProofVerifier.swift
//
//  RoleProofVerifier.swift
//  Crisis Connect
//
//  Created by Assistant on 11.03.2026.
//

import Foundation
import CryptoKit

struct RoleProofVerifier {
    let maxClockSkewMillis: Int64
    let maxExpiredCertificateGraceMillis: Int64
    let timeProvider: () -> Int64

    init(
        maxClockSkewMillis: Int64 = 60_000,
        maxExpiredCertificateGraceMillis: Int64 = 7 * 24 * 60 * 60 * 1000,
        timeProvider: @escaping () -> Int64 = { Int64(Date().timeIntervalSince1970 * 1000) }
    ) {
        self.maxClockSkewMillis = maxClockSkewMillis
        self.maxExpiredCertificateGraceMillis = max(0, maxExpiredCertificateGraceMillis)
        self.timeProvider = timeProvider
    }

    func verifyPacket(
        key: SymmetricKey,
        transportPacket: Data,
        maxPacketSize: Int,
        expectedSessionNonce: String? = nil
    ) throws -> RoleProofPayload {
        let plaintext = try BleAesGcm.decryptPacket(
            key: key,
            transportPacket: transportPacket,
            maxPacketSize: maxPacketSize
        )
        let proof = try RoleProofPayload.fromJSON(plaintext)
        return try verifyProofPayload(
            proof: proof,
            expectedSessionNonce: expectedSessionNonce
        )
    }

    func verifyProofPayload(
        proof: RoleProofPayload,
        expectedSessionNonce: String? = nil
    ) throws -> RoleProofPayload {
        try validate(proof, expectedSessionNonce: expectedSessionNonce)
        return proof
    }

    private func validate(
        _ proof: RoleProofPayload,
        expectedSessionNonce: String?
    ) throws {
        let normalizedExpectedNonce = expectedSessionNonce?
            .trimmingCharacters(in: .whitespacesAndNewlines)
        if let normalizedExpectedNonce, !normalizedExpectedNonce.isEmpty {
            guard proof.sessionNonce == normalizedExpectedNonce else {
                throw RoleProofVerificationError.sessionNonceMismatch
            }
        }

        let now = timeProvider()
        let skew = abs(now - proof.timestamp)
        guard skew <= maxClockSkewMillis else {
            throw RoleProofVerificationError.clockSkew
        }

        let certificateBytes = try decodeBase64Data(proof.certificate, fieldName: "certificate")
        guard let roleCertificate = RoleCertificate.fromStorageBytes(certificateBytes) else {
            throw RoleProofVerificationError.invalidCertificate
        }
        guard roleCertificate.isUsable(
            at: now,
            allowOfflineGrace: true,
            maxClockSkewMillis: maxClockSkewMillis,
            offlineGraceMillis: maxExpiredCertificateGraceMillis
        ) else {
            throw RoleProofVerificationError.expiredCertificate
        }
        guard RoleCertificateSignatureVerifier.verify(
            roleCertificate: roleCertificate,
            publicKeyBase64: proof.devicePublicKey
        ) else {
            throw RoleProofVerificationError.invalidCertificate
        }

        let devicePublicKeyData = try decodeBase64Data(proof.devicePublicKey, fieldName: "devicePublicKey")
        let devicePublicKey = try P256.Signing.PublicKey(
            x963Representation: BleKeyDecoder.x963PublicKey(from: devicePublicKeyData)
        )
        let signature = try P256.Signing.ECDSASignature(
            derRepresentation: decodeBase64Data(proof.signature, fieldName: "signature")
        )
        guard devicePublicKey.isValidSignature(signature, for: proof.signaturePayloadBytes()) else {
            throw RoleProofVerificationError.invalidSignature
        }
    }
}

enum RoleProofVerificationError: Error {
    case sessionNonceMismatch
    case clockSkew
    case invalidCertificate
    case expiredCertificate
    case invalidSignature
}

File: Android/app/src/main/java/com/auralis/crisisconnect/security/RoleProofVerifier.kt
package com.auralis.crisisconnect.security

import android.util.Base64
import android.util.Log
import java.nio.charset.StandardCharsets
import java.security.KeyFactory
import java.security.PublicKey
import java.security.spec.X509EncodedKeySpec
import kotlin.math.abs

/**
 * Device-side helper that verifies the encrypted role proof received via BLE.
 */
class RoleProofVerifier(
    private val maxPacketSize: Int = 1024,
    private val maxClockSkewMillis: Long = 60_000L,
    private val maxExpiredCertificateGraceMillis: Long = 0L,
    private val timeProvider: () -> Long = { System.currentTimeMillis() },
    private val tag: String = "RoleProofVerifier"
) {

    init {
        require(maxExpiredCertificateGraceMillis >= 0L) {
            "maxExpiredCertificateGraceMillis must be non-negative"
        }
    }

    fun verifyPacket(
        keyBytes: ByteArray,
        transportPacket: ByteArray,
        expectedSessionNonce: String? = null
    ): RoleProofVerificationResult {
        return runCatching {
            val encrypted = AesCipherHelper.unwrapTransportPacket(transportPacket, maxPacketSize)
            val plaintext = AesCipherHelper.decrypt(keyBytes, encrypted)
            val jsonString = plaintext.toString(StandardCharsets.UTF_8)
            val proof = RoleProofPayload.fromJson(jsonString)
            verifyProofPayload(proof, expectedSessionNonce = expectedSessionNonce)
        }.getOrElse { throwable ->
            Log.e(tag, "Failed to verify role proof", throwable)
            RoleProofVerificationResult.Failure(throwable.message ?: "Unknown failure", throwable)
        }
    }

    fun verifyProofPayload(
        proof: RoleProofPayload,
        expectedSessionNonce: String? = null
    ): RoleProofVerificationResult {
        return runCatching {
            validateProof(proof, expectedSessionNonce = expectedSessionNonce)
            RoleProofVerificationResult.Success(proof)
        }.getOrElse { throwable ->
            Log.e(tag, "Role proof validation failed", throwable)
            RoleProofVerificationResult.Failure(throwable.message ?: "Unknown failure", throwable)
        }
    }

    private fun validateProof(
        proof: RoleProofPayload,
        expectedSessionNonce: String? = null
    ) {
        val normalizedExpectedNonce = expectedSessionNonce?.trim()?.takeIf { it.isNotEmpty() }
        if (normalizedExpectedNonce != null) {
            require(proof.sessionNonce == normalizedExpectedNonce) {
                "Role proof session nonce mismatch"
            }
        }

        val now = timeProvider()
        val skew = abs(now - proof.timestamp)
        require(skew <= maxClockSkewMillis) {
            "Role proof timestamp skew $skew ms exceeds $maxClockSkewMillis"
        }

        val certificateBytes = decodeBase64(proof.certificate, "certificate")
        val roleCertificate = RoleCertificate.fromStorageBytes(certificateBytes)
            ?: throw IllegalArgumentException("Certificate format is invalid")
        require(roleCertificate.role in RoleCertificate.ALLOWED_ROLES) {
            "Certificate role '${roleCertificate.role}' is not authorized"
        }

        val earliestValidMillis = roleCertificate.issuedAtMillis - maxClockSkewMillis
        require(now >= earliestValidMillis) {
            "Certificate is not yet valid"
        }
        // Enforce offline grace locally; do not trust a client-supplied flag to expand it.
        val latestAllowedWithGrace =
            roleCertificate.expiresAtMillis + maxClockSkewMillis + maxExpiredCertificateGraceMillis
        require(now <= latestAllowedWithGrace) {
            "Certificate expired beyond offline grace window"
        }

        val devicePublicKey = decodeEcPublicKey(proof.devicePublicKey)
        val signatureBytes = decodeBase64(proof.signature, "signature")

        val certificateValid = RoleCertificateSignatureVerifier.verify(
            roleCertificate = roleCertificate,
            publicKeyBase64 = proof.devicePublicKey
        )
        require(certificateValid) { "Certificate validation failed" }

        val signatureValid = Crypto.verifySignature(
            devicePublicKey,
            proof.signaturePayloadBytes(),
            signatureBytes
        )
        require(signatureValid) { "Payload signature is invalid" }
    }

    private fun decodeEcPublicKey(base64: String): PublicKey {
        val decoded = decodeBase64(base64, "devicePublicKey")
        val spec = X509EncodedKeySpec(decoded)
        return KeyFactory.getInstance("EC").generatePublic(spec)
    }

    private fun decodeBase64(value: String, fieldName: String): ByteArray {
        require(value.isNotBlank()) { "$fieldName is required" }
        val decoded = try {
            Base64.decode(value.trim().replace("\n", ""), Base64.NO_WRAP or Base64.NO_PADDING)
        } catch (error: IllegalArgumentException) {
            throw IllegalArgumentException("$fieldName is not valid Base64", error)
        }
        require(decoded.isNotEmpty()) { "$fieldName must not be empty" }
        return decoded
    }
}

sealed class RoleProofVerificationResult {
    data class Success(val proof: RoleProofPayload) : RoleProofVerificationResult()
    data class Failure(val reason: String, val cause: Throwable? = null) : RoleProofVerificationResult()
}
